processing/src/main/java/org/apache/druid/segment/ColumnarValueUnnestCursor.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment;
+
+import org.apache.druid.query.BaseQuery;
+import org.apache.druid.query.dimension.DimensionSpec;
+import org.apache.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import org.apache.druid.segment.column.ColumnCapabilities;
+import org.joda.time.DateTime;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+
+public class ColumnarValueUnnestCursor implements Cursor
+{
+  private final Cursor baseCursor;
+  private final ColumnSelectorFactory baseColumSelectorFactory;
+  private final ColumnValueSelector columnValueSelector;
+  private final String columnName;
+  private final String outputName;
+  private final LinkedHashSet<String> allowSet;
+  private int index;
+  private Object currentVal;
+  private List<Object> unnestListForCurrentRow;
+  private boolean needInitialization;
+
+  public ColumnarValueUnnestCursor(
+      Cursor cursor,
+      ColumnSelectorFactory baseColumSelectorFactory,
+      String columnName,
+      String outputColumnName,
+      LinkedHashSet<String> allowSet
+  )
+  {
+    this.baseCursor = cursor;
+    this.baseColumSelectorFactory = baseColumSelectorFactory;
+    this.columnValueSelector = this.baseColumSelectorFactory.makeColumnValueSelector(columnName);
+    this.columnName = columnName;
+    this.index = 0;
+    this.outputName = outputColumnName;
+    this.needInitialization = true;
+    this.allowSet = allowSet;
+  }
+
+  @Override
+  public ColumnSelectorFactory getColumnSelectorFactory()
+  {
+    return new ColumnSelectorFactory()
+    {
+      @Override
+      public DimensionSelector makeDimensionSelector(DimensionSpec dimensionSpec)
+      {
+        throw new UnsupportedOperationException("Dimension selector not applicable for column value selector");
```

Comment:
Errr, you did too much!  I was only talking about the one case where something asks for the column that is being unnested.  It's totally possible that one of the other columns is being accessed as a DimensionSelector and you want to still allow for that.

Reply:
fixed